File: 8. 기타 그래프 이론/탑승구.js
input_data = `
2
2
3
3
4
4
`.trim().split('\n')

const [G, P] = [4, 6] // 탑승구 수, 비행기 수
const parent = new Array(G+1).fill().map((val, idx) => idx)
let ans = 0
for (let data of input_data) {
  g = find_parent(parseInt(data))
  
  if (g === 0) {
    break
  } else {
    union_parent(g, g-1)
    ans += 1
  }
}
console.log(ans)


function find_parent(x) {
  if (x != parent[x]) {
    parent[x] = find_parent(parent[x])
  }
  return parent[x]
}

function union_parent(a, b) {
  a = find_parent(a)
  b = find_parent(b)
  if (a > b) {
    parent[a] = b
  } else {
    parent[b] = a
  }
}

File: 8. 기타 그래프 이론/탑승구.py
def find_parent(x):
    if parent[x] != x:
        parent[x] = find_parent(parent[x])
    return parent[x]

def union_parent(a, b):
    a = find_parent(a)
    b = find_parent(b)
    if a > b:
        parent[a] = b
    else:
        parent[b] = a

G = int(input()) # 탑승구 수
P = int(input()) # 비행기 수

parent = [i for i in range(G+1)]
ans = 0
for _ in range(P):
  g = find_parent(int(input()))
  
  if not g:
    break
  else:
    union_parent(g, g-1)
    ans += 1

print(ans)